types/underscore/index.d.ts

```diff
@@ -562,36 +518,27 @@ declare module _ {
         * @param arguments Additional arguments to pass to the method `methodName`.
         **/
         invoke<T extends {}>(
```

Comment:
Having just tried out `_.map([{a: 'a'}, null, undefined], 'a')`, `_.invoke([{a: 'a'}, null, undefined], 'toString')`, and `_.pluck([{a: 'a'}, undefined, null], 'a')`, it looks like Underscore handles undefined and null values for such iterators and returns `undefined` when they're encountered, so maybe I should update return types for such functions that take items that may be null or undefined to include undefined results.

Reply:
What I was going to do here was

```TS
// underscore seems to propagate both null and undefined to undefined
type NullPropagating<T, TResult> = T extends undefined | null ? TResult | undefined : TResult;

map<T, TProperty extends keyof NonNullable<T>>(
            collection: _.Collection<T>,
            iterator: TProperty): NullPropagating<T, NonNullable<T>[TProperty]>[];
```

but this actually brings up an interesting case; for example, if I have the below

```TS
const array: ({ a: string } | { b: string } | undefined)[] = [{ a: 'a' }, { b: 'b' }, undefined];
const result = _.map(array, 'a');
```

the result will be `['a', undefined, undefined]`...so what I really want the definition here to be is something like

```TS
map<T, TProperty extends string>(
            collection: _.Collection<T>,
            iterator: TProperty): (TProperty in T ? T[TProperty] : undefined)[];
```

but I don't think that's a thing I can do at the moment unfortunately (though possibly I just don't know the syntax for it).

Reply:
[This SO article](https://stackoverflow.com/questions/49401866/all-possible-keys-of-an-union-type) provides a sneaky way to constrain the key to a union of all the keys for the set of types in a union type T, but even doing the below to try and pull the type of the specified property from each type in T separately still doesn't quite seem to work...

```TS
type KeysOfUnion<T> = T extends any ? keyof T : never;

map<T, TProperty extends KeysOfUnion<T>>(
            collection: _.Collection<T>,
            iterator: TProperty): (T extends { TProperty: infer TResult } ? TResult : undefined)[];
```

Reply:
Huzzah, a thing that works here! Now I've just got to figure out if there's a way to simplify it at all because it's a bit of an ugly-looking mess.

```TS
type KeysOfUnion<T> = T extends any ? keyof T : never;

type TypesOfUnion<T, TProperty extends KeysOfUnion<T>> = T extends any ? T[Extract<keyof T, TProperty>] extends never ? undefined : T[Extract<keyof T, TProperty>] : never;

map<T, TProperty extends KeysOfUnion<T>>(
            collection: _.Collection<T>,
            iterator: TProperty): TypesOfUnion<T, TProperty>[];

const array: ({ a: string } | undefined)[] = [{ a: 'a' }, { a: 'b' }];
const property = 'a';
let result: (string | undefined)[];

result = _.map<{ a: string } | undefined, typeof property>(array, property); // return type matches expected type!
```